src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -554,25 +563,84 @@ private async Task InitializeAsync(bool ignoreFailures, IEnumerable<Configuratio
+
+                                    watchedSettings = UpdateWatchedKeyValueCollections(watchedSettings, data);
+                                },
+                                cancellationToken)
+                                .ConfigureAwait(false);
+
+                            loadSuccess = true;
+                        }
+                        // Tries to throw the last exception before startup timed out to give information on request failure instead of just
+                        // OperationCanceledException when startup times out.
+                        catch (RequestFailedException exception)
+                        {
+                            if (IsFailOverable(exception))
+                            {
+                                previousException = exception;
+                            }
+                            else
+                            {
+                                throw;
+                            }
+                        }
+                        catch (Exception exception) when ((exception as AggregateException)?.InnerExceptions?.Any(e =>
+                            e is RequestFailedException) ?? false)
+                        {
+                            if (IsFailOverable(exception as AggregateException))
+                            {
+                                previousException = exception;
+                            }
+                            else
+                            {
+                                throw;
+                            }
+                        }
+                        catch (Exception exception) when (exception is OperationCanceledException ||
+                            ((exception as AggregateException)?.InnerExceptions?.All(e =>
+                            e is OperationCanceledException) ?? false))
+                        {
+                            if (previousException != null)
+                            {
+                                throw previousException;
+                            }
+
+                            throw;
+                        }
+                    }
+
+                    if (!loadSuccess)
                     {
-                        data = await LoadSelectedKeyValues(
-                            client,
-                            cancellationToken)
-                            .ConfigureAwait(false);
-
-                        watchedSettings = await LoadKeyValuesRegisteredForRefresh(
-                            client,
-                            data,
-                            cancellationToken)
-                            .ConfigureAwait(false);
-
-                        watchedSettings = UpdateWatchedKeyValueCollections(watchedSettings, data);
-                    },
-                    cancellationToken)
-                    .ConfigureAwait(false);
+                        await Task.Delay(1000).ConfigureAwait(false);
```

Comment:
Can we make this 1s a const? This is the minimum delay between retries. Not sure if we should have a maximal delay and slowly grow the delay between retries. Thoughts?

Reply:
The actual server calls are already retried with some exponential backoff strategy. I dont see why we would need to slow down the overall startup retries.

Reply:
Unless the concern is that this can be misused by customers if they set SDK retries to 0. If SDK is not retrying, we'll keep retrying every 1 second which is not good. So having this second layer of retries with exponential backoff might be good.

Reply:
Imagining there is only one client, we don't want to beat that client in a tight loop.

Reply:
I think Amer will be reworking this a bit. I agree that we should backoff on the top level. I suggested a flow like below.

```
public async Task LoadAsync()
{
    var cts = new CancellationTokenSource(_startupOptions.Timeout);

    var errors = new List<Exception>();

    int attempts = 0;

    while (true)
    {
        attempts++;

        IEnumerable<ConfigurationClient> clients = _clientManager.GetClients(ignoreBackoff: true);

        if (await TryInitializeAsync(clients, errors, cts.Token))
        {
            break;
        }

        try
        {
            await Task.Delay(Timespan.FromSeconds(1).Backoff(attempts), cts.Token);
        }
        catch (OperationCanceledException)
        {
            throw new TimeoutException(
                new AggregateException(errors));
        }
    }
}

//
// TryInitializeAsync
// It swallows (returns false for) operation canceled exception
// It swallows failoverable exceptions
// It propagates non-failoverable exceptions
// It fills the 'errors' list with swallowed exceptions
```